clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -128,11 +128,12 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                     long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                     this.waitTime.record(timeNs, time.milliseconds());
 
-                    if (waitingTimeElapsed)
+                    if (waitingTimeElapsed) {
+                        this.waiters.remove(moreMemory);
```

Comment:
@MayureshGharat Should we also make sure that, in the event the thread is interruptted when it is blocked on `boolean waitingTimeElapsed = !moreMemory.await(remainingTimeToBlockNs, TimeUnit.NANOSECONDS)`, the condition is removed from queue before `InterruptedException` is thrown?